src/os: implement os.File.Fd() method

ZauberNerd opened the pull request:
This commits adds a OS-specific `Fd()` function for `file_anyos.go` and
`file_other.go`, which returns a uintptr to the underlying file
descriptor.

Creating this as a draft PR, because I'm unsure, if this is the correct way to go about implementing it.

ZauberNerd committed:
src/os: implement os.File.Fd() method

This commits adds a OS-specific `Fd()` function for `file_anyos.go` and
`file_other.go`, which returns a uintptr to the underlying file
descriptor.

dgryski commented:
Thanks.  Can you add a test or two?

dkegel-fastly commented:
It looks ok.  At least a trivial test would be nice.  
(Upstream barely tests it at all, using it to force blocking mode (!) and as a probe for an assertion in https://go.dev/src/os/readfrom_linux_test.go.  If we pass that test I'd be impressed.)

Every time we fatten the FileHandle interface, it diverges from https://pkg.go.dev/tinygo.org/x/tinyfs#FileHandle, but I don't know whether that breaks users of tinyfs somehow.

dgryski commented:
Should we be having PRs that update tinyfs at all?

/cc @aykevl

ZauberNerd commented:
> It looks ok. At least a trivial test would be nice. (Upstream barely tests it at all, using it to force blocking mode (!) and as a probe for an assertion in https://go.dev/src/os/readfrom_linux_test.go. If we pass that test I'd be impressed.)

I'm thinking maybe it's possible to `os.Open("/proc/self/<something>")` to read the file descriptor of the just opened file?
I know about `/proc/self/fd/*`, but for that I would need to know the descriptor already.

ZauberNerd commented:
Otherwise I can at least test that `Stdin.Fd() == 0`, `Stdout.Fd() == 1` etc.

dkegel-fastly commented:
I... guess you could pass the fd to NewFile and make sure you can use that file?

(cf. upstream's TestNewFileBlock(), which passes it an fd from a syscall and makes sure it can do file.Read().)

dkegel-fastly commented:
Offhand I would avoid /proc for this test.

ZauberNerd committed:
test: simple test to verify os.File.Fd() is working

ZauberNerd committed:
test: write into a temp file and read from its fd

This test creates a new temp file and writes some bytes into it.
It then opens a new file for the file descriptor of the temp file and
tries to read some bytes out of it.

ZauberNerd commented:
@dkegel-fastly  I'm not sure, if I understood you correctly, but I tried to write some meaningful tests - let me know if there's anything I should change or if I can mark this PR as ready for review.

dkegel-fastly commented:
Looks like it's unhappy on windows; if you have wine, you can reproduce it without windows, e.g.
```
$ GOOS=windows tinygo test -c os
$ wine os.test -test.v -test.run TestFd
=== RUN   TestFd
--- FAIL: TestFd (0.00s)
    Failed to read 5 bytes from file descriptor: readat as-fd: operation not implemented
    File descriptor contents not equal to file contents.
FAIL
```
This is not surprising, as fds are very different on windows, and upstream's "pollfd" layer that hides the difference (I think) has not been ported yet.

So just skip that test on windows, as TestMkdirTempBadDir() does.

deadprogram commented:
Please rebase against the latest dev, so that the Windows tests can run correctly. Thanks!

dkegel-fastly reviewed (approved):
LGTM

aykevl commented on src/os/filesystem.go:
```diff
@@ -57,6 +57,8 @@ type FileHandle interface {
 
 	// Close closes the file, making it unusable for further writes.
 	Close() (err error)
+
+	Fd() uintptr
```
Instead of adding a new method to this interface (which breaks packages like tinyfs!), I would suggest type-asserting in the `Fd()` function.

dkegel-fastly replied:
tinyfs was probably broken when we added the Seek() and ReadAt() methods to FileHandle, but yeah, avoiding further divergence would probably be good.

dkegel-fastly replied:
See also https://github.com/tinygo-org/tinygo/issues/2709

dkegel-fastly replied:
Tinyfs itself doesn't care about the discrepancy, its tests still pass because it has its own copy of File and FileHandle, and doesn't use tinygo's...?

ZauberNerd replied:
Updated to the suggestion from @aykevl.

aykevl commented on src/os/file.go:
```diff
@@ -181,7 +181,7 @@ func (f *File) SyscallConn() (syscall.RawConn, error) {
 // Fd returns the file handle referencing the open file.
 func (f *File) Fd() uintptr {
-	panic("unimplemented: os.file.Fd()")
+	return f.handle.Fd()
```
Here is an example:

```suggestion
	handle, ok := f.handle.(interface {
		Fd() uintptr
	})
	if ok {
		return handle.Fd()
	}
	return 0
```

(This inline interface is perhaps a bit ugly though)

deadprogram commented:
Now merging. Thanks @ZauberNerd
